test/e2e/konnect_test.go

```diff
@@ -190,28 +192,113 @@ func createTestRuntimeGroup(ctx context.Context, t *testing.T) string {
-		Labels:      &rg.Labels{"created_in_tests": "true"},
-		Name:        uuid.NewString(),
-		ClusterType: rg.ClusterTypeKubernetesIngressController,
-	})
-	require.NoError(t, err, "failed to create runtime group")
-	require.Equalf(t, http.StatusCreated, createRgResp.StatusCode(), "failed creating RG: %s", string(createRgResp.Body))
-	require.NotNil(t, createRgResp.JSON201)
-	require.NotNil(t, createRgResp.JSON201.Id)
-	id := *createRgResp.JSON201.Id
 	t.Cleanup(func() {
-		_, err := rgClient.DeleteRuntimeGroupWithResponse(ctx, id)
-		assert.NoErrorf(t, err, "failed to cleanup a runtime group: %q", id)
+		// delete the RG used in the test.
+		t.Logf("deleting runtime group %s", rgID)
+		err := retry.Do(
+			func() error {
+				_, err := rgClient.DeleteRuntimeGroupWithResponse(ctx, rgID)
+				if err != nil {
+					t.Logf("failed to delete runtime group %s: %v", rgID.String(), err)
+					return err
+				}
+				return nil
+			},
+			retry.Attempts(5), retry.Delay(time.Second),
+		)
+		assert.NoErrorf(t, err, "failed to delete runtime group %s used in the test", rgID.String())
+
+		// list other RGs also created in the previous code for situations
+		// that RG created, but we timed out on getting the response.
+		// This prevents RG leaks.
+		t.Logf("deleting other runtime groups created in the test")
+		rgNumberLimit := 50
+		err = retry.Do(
+			func() error {
+				listRgResp, err := rgClient.ListRuntimeGroupsWithResponse(ctx, &rg.ListRuntimeGroupsParams{
+					PageSize: lo.ToPtr(rgNumberLimit),
+				})
+				if err != nil {
+					t.Logf("failed to list runtime groups: %v", err)
+					return err
+				}
+				if listRgResp.StatusCode() != http.StatusOK {
+					t.Logf("failed to list runtime groups: code %d, body %s",
+						listRgResp.StatusCode(), string(listRgResp.Body))
+					return fmt.Errorf("non-OK status code: %d", listRgResp.StatusCode())
+				}
+				if listRgResp.JSON200.Data == nil {
+					t.Logf("failed to list runtime groups: no data in body")
+					return errors.New("no data in response body")
+				}
+
+				rgs := *listRgResp.JSON200.Data
+
+				var deleteRgErrors []error
+				for _, rg := range rgs {
+					if rg.Labels == nil || *rg.Labels == nil {
+						continue
+					}
+					labels := *rg.Labels
+					// find the RGs created in the test.
+					if labels["created_in_tests"] == "true" &&
+						labels["create_test_timestamp"] == createTimeStampStr &&
+						labels["create_test_uuid"] == createTestUUID &&
+						rg.Id != nil {
+						deleteRgID := *rg.Id
+
+						t.Logf("deleting runtime group %s created in test %s", deleteRgID, t.Name())
+						_, deleteErr := rgClient.DeleteRuntimeGroupWithResponse(ctx, deleteRgID)
+						if deleteErr != nil {
+							t.Logf("failed to delete runtime group %s: %v", deleteRgID, err)
+							deleteRgErrors = append(deleteRgErrors, deleteErr)
+						}
+					}
+				}
+				return errors.Join(deleteRgErrors...)
+			},
+			retry.Attempts(5), retry.Delay(time.Second),
```

Comment:
This whole section responsible for deleting RGs could be a function to increase the  readability of this section. WDYT?

Reply:
extracted the procedure to another function.